tests/e2e/runtime.go

```diff
@@ -0,0 +1,465 @@
+			CommitInterval:  epochInterval / 2,
+			RevealInterval:  (epochInterval / 2) - 4,
+			TransitionDelay: 4,
+		}
+	}
+
+	return ff, nil
+}
+
+func (sc *runtimeImpl) resolveRuntimeBinaries(runtimeBinaries []string) map[node.TEEHardware][]string {
+	binaries := make(map[node.TEEHardware][]string)
+	for _, tee := range []node.TEEHardware{
+		node.TEEHardwareInvalid,
+		node.TEEHardwareIntelSGX,
+	} {
+		for _, binary := range runtimeBinaries {
+			binaries[tee] = append(binaries[tee], sc.resolveRuntimeBinary(binary))
+		}
+	}
+	return binaries
+}
+
+func (sc *runtimeImpl) resolveRuntimeBinary(runtimeBinary string) string {
+	path, _ := sc.Flags.GetString(cfgRuntimeBinaryDirDefault)
+	return filepath.Join(path, runtimeBinary)
+}
+
+func (sc *runtimeImpl) waitNodesSynced() error {
+	ctx := context.Background()
+
+	checkSynced := func(n *oasis.Node) error {
+		c, err := oasis.NewController(n.SocketPath())
+		if err != nil {
+			return fmt.Errorf("failed to create node controller: %w", err)
+		}
+		defer c.Close()
+
+		if err = c.WaitSync(ctx); err != nil {
+			return fmt.Errorf("failed to wait for node to sync: %w", err)
+		}
+		return nil
+	}
+
+	sc.Logger.Info("waiting for all nodes to sync")
+
+	for _, n := range sc.Net.Validators() {
+		if err := checkSynced(&n.Node); err != nil {
+			return err
+		}
+	}
+	for _, n := range sc.Net.StorageWorkers() {
+		if err := checkSynced(&n.Node); err != nil {
+			return err
+		}
+	}
+	for _, n := range sc.Net.ComputeWorkers() {
+		if err := checkSynced(&n.Node); err != nil {
+			return err
+		}
+	}
+	for _, n := range sc.Net.Clients() {
+		if err := checkSynced(&n.Node); err != nil {
+			return err
+		}
+	}
+
+	sc.Logger.Info("nodes synced")
+	return nil
+}
+
+func (sc *runtimeImpl) Run(childEnv *env.Env) error {
+	// Start the test network.
+	if err := sc.Net.Start(); err != nil {
+		return err
+	}
+
+	// Wait for all nodes to sync.
+	if err := sc.waitNodesSynced(); err != nil {
+		return err
+	}
+
+	// Connect to the client node.
+	clients := sc.Net.Clients()
+	if len(clients) == 0 {
+		return fmt.Errorf("client initialization failed")
+	}
+
+	conn, err := grpc.Dial(clients[0].SocketPath(),
+		grpc.WithDefaultCallOptions(grpc.ForceCodec(&cmnGrpc.CBORCodec{})),
+		grpc.WithInsecure(),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}))
+	if err != nil {
+		return err
+	}
+	rtc := client.New(conn, runtimeID)
+
+	// Run the given tests for this runtime.
+	for _, test := range sc.RunTest {
+		testName := runtime.FuncForPC(reflect.ValueOf(test).Pointer()).Name()
+
+		sc.Logger.Info("running test", "test", testName)
+		if testErr := test(sc.Net, rtc); testErr != nil {
+			sc.Logger.Error("test failed",
+				"test", testName,
+				"err", testErr,
+			)
+			return testErr
+		}
+		sc.Logger.Info("test passed", "test", testName)
+	}
+
+	return sc.Net.CheckLogWatchers()
+}
+
+// Ugh, this is really annoying :/
```

Comment:
Why are you doing it like that? There are provisions for doing this in the SDK. See an example here (that also uses predefined test keys):
https://github.com/oasisprotocol/oasis-bridge/blob/master/examples/user-witness-flow/main.go

Reply:
Since the SDK doesn't have any docs yet, it's not entirely obvious (at least to me) how things should be done.
Thanks for the link, I'll update the code to use the test keys :)